### PR TITLE
Multi-image masking for single IP Adapter

### PR DESCRIPTION
**Pull request by fabiorigano**
# What does this PR do?
Image masking is currently supported by passing a mask and an image for each IP Adapter in the pipeline.

This PR extends this usage by allowing multiple masks and images to be passed to a single IP Adapter. Scales must be passed accordingly.

Discussed in #7419 

@sayakpaul @yiyixuxu

### CODE REVIEW

File: src/diffusers/models/attention_processor.py [hunk 1]
```diff
@@ -2372,11 +2372,11 @@ def __call__(
...
-            if not isinstance(ip_adapter_masks, torch.Tensor) or ip_adapter_masks.ndim != 4:
-                raise ValueError(
-                    " ip_adapter_mask should be a tensor with shape [num_ip_adapter, 1, height, width]."
-                    " Please use `IPAdapterMaskProcessor` to preprocess your mask"
-                )
+            # if not isinstance(ip_adapter_masks, torch.Tensor) or ip_adapter_masks.ndim != 4:
+                # raise ValueError(
+                    # " ip_adapter_mask should be a tensor with shape [num_ip_adapter, 1, height, width]."
+                    # " Please use `IPAdapterMaskProcessor` to preprocess your mask"
+                # )
```
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:2379]: Can safely go now.

**Review by sayakpaul — COMMENTED** [review]
Changes looking nice ❤️ Great work as always. Let's add a fast test and see how it goes?

**Comment by fabiorigano** [comment]
thank you @sayakpaul 

ready

File: src/diffusers/models/attention_processor.py [hunk 2]
```diff
@@ -2198,14 +2198,10 @@ def __call__(
...
         if ip_adapter_masks is not None:
-            if not isinstance(ip_adapter_masks, torch.Tensor) or ip_adapter_masks.ndim != 4:
-                raise ValueError(
-                    " ip_adapter_mask should be a tensor with shape [num_ip_adapter, 1, height, width]."
-                    " Please use `IPAdapterMaskProcessor` to preprocess your mask"
-                )
             if len(ip_adapter_masks) != len(self.scale):
                 raise ValueError(
-                    f"Number of ip_adapter_masks ({len(ip_adapter_masks)}) must match number of IP-Adapters ({len(self.scale)})"
+                    f"Lenght of ip_adapter_masks array ({len(ip_adapter_masks)}) must match "
```
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:2203]:
```suggestion
                    f"Length of ip_adapter_masks array ({len(ip_adapter_masks)}) must match "
```

File: tests/pipelines/ip_adapters/test_ip_adapter_stable_diffusion.py [hunk 3]
```diff
@@ -544,3 +544,33 @@ def test_ip_adapter_multiple_masks(self):
 
         max_diff = numpy_cosine_similarity_distance(image_slice, expected_slice)
         assert max_diff < 5e-4
+
+    def test_ip_adapter_multiple_masks_one_adapter(self):
```
Comment (sayakpaul) [inline tests/pipelines/ip_adapters/test_ip_adapter_stable_diffusion.py:548]: This is slow test. I was wondering if it's possible to add a fast test like

https://github.com/huggingface/diffusers/blob/c2e87869be81bef9c87adb34b7815adebe042e64/tests/pipelines/test_pipelines_common.py#L238

?

File: src/diffusers/models/attention_processor.py [hunk 4]
```diff
@@ -2198,14 +2198,10 @@ def __call__(
         hidden_states = attn.batch_to_head_dim(hidden_states)
 
         if ip_adapter_masks is not None:
```
[the remark this answers is not in the document]
Comment (yiyixuxu) [reply src/diffusers/models/attention_processor.py:2200]: i put this really long check here
```py
if not (len(ip_adapter_masks) == len(self.scale) == len(ip_hidden_states) == len(self.to_k_ip) == len(self.to_v_ip)
``` 
but if we have already checked any of these somewhere else, you can remove them from the equation 
e.g if `len(self.to_k_ip) == len(self.to_v_ip)` is always true, we don't have to include it here

**Review by yiyixuxu — COMMENTED** [review]
thank you for working on this!
looks good. I left some feedbacks. let's make very clear what's the expected shape for `ip_adapter_mask` and `scale`

File: src/diffusers/models/attention_processor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fabiorigano) [reply src/diffusers/models/attention_processor.py:2200]: > we should still allow them to pass mask as a tensor, no?

yes, I was looking for the best way to have backward compatibility, thanks for the addition

**Comment by fabiorigano** [comment]
@yiyixuxu @sayakpaul now it's ready
thanks for your reviews

**Review by yiyixuxu — APPROVED** [review]
looks good to me! 
I will run slow tests to make sure it does not break anything before merge

**Comment by fabiorigano** [comment]
thank you YiYi!

I fixed the fast test for the panorama pipeline that was failing and the slow test

I ran all the slow tests related to image masking locally and they didn't break :)

**Pull request merged**